File: Data_Structure/python/ch3_Linked_list/LinkedList.py
class LinkedListNode:
    def LinkedListNode(self):
        self.data = 0
        self.pLink = {'HeaderClass':0, 'CurrentClass': 0, 'NextClass':0 }
        self.pLink['CurrentClass'] = self
        self.currentCount = 0

def LinkedList(Node):
    LinkedListNode()
    Node.pLink['HeaderClass'] = Node
    return Node

def createLinkedList(Node):
    Node = LinkedListNode()
    Node.LinkedListNode()
    Node.data = 'HeaderNode'
    LinkedList(Node)
    return Node

def getLinkedListData(pList, position):
    i = 0
    pCurrentNode = LinkedListNode()
    pCurrentNode.LinkedListNode()
    pCurrentNode = pList.pLink['HeaderClass']

    if position > pList.currentCount :
        print('\n불러올 위치 값이 현재 노드 개수 보다 많습니다.\n')
        return

    for i in range(position):
        pCurrentNode = pCurrentNode.pLink['NextClass']

    return pCurrentNode.data

def addLinkedListData(pList, position, data):
    i = 0

    if position == 0 :
        print('\n값을 넣을 위치는 헤더노드의 위치 입니다.\n')

    pNewNode = LinkedListNode()
    pNewNode.LinkedListNode()
    pPreNode = LinkedListNode()
    pPreNode.LinkedListNode()

    pNewNode.data = data

    pPreNode = pList.pLink['HeaderClass']
    for i in range(position-1) :
        pPreNode = pPreNode.pLink['NextClass']

    pNewNode.pLink['NextClass'] = pPreNode.pLink['NextClass']
    pPreNode.pLink['NextClass'] = pNewNode

    pList.currentCount += 1

    return pList

def removeLinkedListData(pList, position):
    if position == 0 :
        print('\n헤더노드는 지울 수 없습니다.\n')
        return
    i = int(0)
    pDelNode = LinkedListNode()
    pDelNode.LinkedListNode()
    pPreNode = LinkedListNode()
    pPreNode.LinkedListNode()

    pPreNode = pList.pLink['HeaderClass']

    for i in range(position - 1):
        pPreNode = pPreNode.pLink['NextClass']

    pDelNode = pPreNode.pLink['NextClass']
    pPreNode.pLink['NextClass'] = pDelNode.pLink['NextClass']

    pList.currentCount -= 1

def deleteLinkedList(pList):
    if pList.pLink == {} :
        print('\n리스트가 비어있습니다.\n')
        return

    pDelNode = LinkedListNode()
    pDelNode.LinkedListNode()
    pPreNode = LinkedListNode()
    pPreNode.LinkedListNode()

    pPreNode = pList.pLink['HeaderClass']

    while(pPreNode.pLink['NextClass'] != 0 ):
        pDelNode = pPreNode
        pPreNode = pPreNode.pLink['NextClass']

        pDelNode.data = 0
        pDelNode.pLink.clear()

def getLinkedListLength(pList):
       return pList.pLink['HeaderClass'].currentCount

def displayList(pList):
    i = 0
    print('\n일반 순회')
    for i in range(pList.currentCount):
        print("Node[{}] = {}".format(i+1,getLinkedListData(pList,i+1)))

def iterateLinkedList(pList):
    count = int(0)
    pNode = LinkedListNode()
    pNode.LinkedListNode()

    if pList.pLink == {} :
        print('\n리스트가 비어있습니다.\n')
        return

    pNode = pList.pLink['HeaderClass']

    print('\n로직 처리 순회')

    while  pNode != 0:
        print('{} = {}'.format(count, pNode.data))
        count += 1
        pNode = pNode.pLink['NextClass']

    print('노드 개수 : {}'.format(getLinkedListLength(pList.pLink['HeaderClass'])))

def concatLinkedList(pListA, pListB):
    pNodeA = LinkedListNode()
    pNodeA.LinkedListNode()
    pNodeB = LinkedListNode()
    pNodeB.LinkedListNode()

    if pListA.pLink['HeaderClass'] != 0 and pListB.pLink['HeaderClass'] != 0:
        pNodeA = pListA.pLink['HeaderClass']
        pNodeA = pNodeA.pLink['NextClass']

        pNodeB = pListB.pLink['HeaderClass']
        pNodeB = pNodeB.pLink['NextClass']

        while pNodeA.pLink['NextClass'] != 0 :
            pNodeA = pNodeA.pLink['NextClass']

        pListA.pLink['HeaderClass'].currentCount = getLinkedListLength(pListA.pLink['HeaderClass']) + getLinkedListLength(pListB.pLink['HeaderClass'])
        pNodeA.pLink['NextClass'] = pNodeB.pLink['CurrentClass']
        pListB.pLink.clear()


pList = LinkedListNode()
pList.LinkedListNode()
pList = createLinkedList(pList)
addLinkedListData(pList, 1, 10)
addLinkedListData(pList, 2, 20)
addLinkedListData(pList, 2, 30)

value = getLinkedListData(pList, 1)
print('위치 : {}\n값 : {}\n'.format(1,value))

displayList(pList)

iterateLinkedList(pList)

removeLinkedListData(pList, 1)

iterateLinkedList(pList)
deleteLinkedList(pList)

pListA = LinkedListNode()
pListA.LinkedListNode()
pListA = createLinkedList(pListA)
pListB = LinkedListNode()
pListB.LinkedListNode()
pListB = createLinkedList(pListB)

addLinkedListData(pListA, 1, 10)
addLinkedListData(pListA, 2, 20)
addLinkedListData(pListA, 3, 30)

addLinkedListData(pListB, 1, 40)
addLinkedListData(pListB, 2, 50)

print('연결 리스트 결합 전')
print('\n연결 리스트A')
iterateLinkedList(pListA)
print('\n연결 리스트B')
iterateLinkedList(pListB)

concatLinkedList(pListA, pListB)
print('\n연결 리스트 결합 후')
print('\n연결 리스트A')
iterateLinkedList(pListA)
print('\n연결 리스트B')
iterateLinkedList(pListB)

deleteLinkedList(pListA)
deleteLinkedList(pListB)

File: Data_Structure/Cprogramming/ch11_Graph/linkedlist.c
#include "linkedlist.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

LinkedList *createLinkedList() {
	LinkedList *pReturn = (LinkedList*)malloc(sizeof(LinkedList));
	memset(pReturn, 0, sizeof(LinkedList));   // 새로 메모리를 할당한 다음에 0으로 초기화 ( 버그를 없애기 위함 )
	return pReturn;
}   // 리스트 생성

int getLinkedListData(LinkedList* pList, int position) {
	int i = 0;

	LinkedListNode *pCurrentNode = &(pList->headerNode);  // pCurrentNode라는 구조체를 만들어주고 파라미타로 받은 리스트의 헤더 주소를 넣는다.
	for (i = 0; i <= position; i++) {      // position +1 횟수만큼 노드의 링크를 이용하여 다음 노드로 이동한다.
		pCurrentNode = pCurrentNode->pLink;
	}

	return pCurrentNode->data;
}   // 리스트의 값 불러오기

int addLinkedListData(LinkedList* pList, int position, int data) {
	int i = 0;
	LinkedListNode *pNewNode = NULL;
	LinkedListNode *pPreNode = NULL;

	pNewNode = (LinkedListNode*)malloc(sizeof(LinkedListNode));
	pNewNode->data = data;

	pPreNode = &(pList->headerNode);    // pPreNode에 리스트의 헤더노드 주소를 가져온다.
	for (i = 0; i < position; i++) {
		pPreNode = pPreNode->pLink;
	}   // 새로 자료를 추가하려는 위치의 이전 노드까지 이동 시킨다.

	pNewNode->pLink = pPreNode->pLink;  // 새로운 노드의 링크를 이전의 노드가 가르키던 다음의 노드의 링크를 가르키게 한다.
	pPreNode->pLink = pNewNode;     // 이전의 노드의 링크를 새로운 노드를 가르키게 한다.
	pList->currentCount++;  // 현재 노드의 개수를 1 증가 시킨다.
	return 0;
}   // 새로운 노드 생성

int removeLinkedListData(LinkedList* pList, int position) {
	int i = 0;
	LinkedListNode *pDelNode = NULL;
	LinkedListNode *pPreNode = NULL;

	pPreNode = &(pList->headerNode);
	for (i = 0; i < position; i++) {
		pPreNode = pPreNode->pLink;
	} // 제거하려는 노드의 이전 노드까지 이동

	pDelNode = pPreNode->pLink;   // 제거하려는 노드 지정 ( 지정된 pPreNode의 다음 링크가 제거하려는 노드이다.)
	pPreNode->pLink = pDelNode->pLink;  // 이전 노드의 처리 ( 이전 노드의 링크를 지울려는 링크의 다음 링크를 가르키게 한다.)
	free(pDelNode); // 메모리 해제
	pList->currentCount--; // 현재 노드 개수 1개 감소
	return 0;
}   // 기존 노드 제거

void deleteLinkedList(LinkedList* pList) {
	LinkedListNode *pDelNode = NULL;
	LinkedListNode *pPreNode = pList->headerNode.pLink;
	while (pPreNode != NULL) {
		pDelNode = pPreNode;
		pPreNode = pPreNode->pLink;

		free(pDelNode);
	}   // 각노드를 처음부터 순서대로 메모리 해체 시킨다.

	free(pList);    // 연결 리스트 자체를 메모리 해체 시킨다.
}   // 리스트 제거

int getLinkedListLength(LinkedList* pList) {
	if (NULL != pList) {
		return pList->currentCount;
	}
	return 0;
}   // 리스트의 길이 호출

void displayList(LinkedList *pList) {
	int i = 0;
	printf("일반 순회\n");
	for (i = 0; i < pList->currentCount; i++) {
		printf("[%d] = %d\n", i, getLinkedListData(pList, i));
	}
	printf("\n");
}   //  순회 : 연결 리스트의 노드를 차례대로 방문하는 것. O(n^2) 의 시간

void iterateLinkedList(LinkedList* pList)
{
	int count = 0;
	LinkedListNode* pNode = NULL;

	pNode = pList->headerNode.pLink;

	printf("로직 처리 순회\n");
	while (pNode != NULL) {
		printf("[%d] = %d\n", count, pNode->data);  // 로직 처리 부분 : 헤더 노드의 링크를 가져와서 while 루프에서 데이터를 처리한다.
		count++;
		pNode = pNode->pLink;
	}
	printf("노드 개수 : %d\n", count);

}   //  루프 내에서 로직을 처리하도록 한 순회

void concatLinkedList(LinkedList* pListA, LinkedList* pListB) {
	LinkedListNode* pNodeA = NULL;   // 연결 리스트 pListA의 마지막 노드를 가리킬 pNodeA

	if (pListA != NULL && pListB != NULL) {   // 입력 파라미터 유효성 점검
		pNodeA = pListA->headerNode.pLink;    // pNodeA가 연결리스트 pListA의 첫 번째 노드를 가리키게 한다.

		while (pNodeA != NULL && pNodeA->pLink != NULL) { // pNodeA가 마지막 노드를 가리킬 때까지 루프를 돈다.
			pNodeA = pNodeA->pLink;
		}
		pNodeA->pLink = pListB->headerNode.pLink;   // pNodeA의 다음 노드로 pListB의 첫 번째 노드를 설정한다.
		pListB->headerNode.pLink = NULL;    // 연결 리스트 pListB는 빈 연결 리스트가 되도록 한다.
	}
}   // 리스트와 리스트를 연결해주는 함수